### ISSUE TITLE
Sponsored Smart Contract transactions fail

### ISSUE THREAD

**Issue by alihamze** [stx-labs/stacks.js#1122]
**Describe the bug**
Attempting to broadcast a sponsored smart contract call results in the following error:
```json
{"error":"transaction rejected","reason":"SignatureValidation","reason_data":{"message":"Invalid signature: failed to recover public key"},"txid":"6def108f5240757c9e8e3d8ade88c5c26ff31a1f1f4533cb756ba9e8410e5d00"}
```
Broadcasting the same transaction without a sponsor is successfully processed.

**What version of Blockstack.js are you using?**
@stacks/blockchain-api-client@0.34.2
@stacks/transactions@2.0.1

**Is the bug present in the Blockstack CLI, Gaia hub, Connect or Blockstack Browser?**
N/A

**To Reproduce**
The call is made using the format from the [README](https://github.com/blockstack/stacks.js/blob/8b4fdb5bdf5e14f80e8c8169d3d281a6940ba087/packages/transactions/README.md#sponsoring-transactions):

```js
const contractAddress = "";
const contractName = "";
const userPrivateKey = "";
const sponsorPrivateKey = "";
const stacksNetwork = new StacksTestnet();
const anchorMode = AnchorMode.Any;

const txOptions = {
    contractAddress: contractAddress,
    contractName: contractName,
    functionName: 'mint-nft',
    functionArgs: [
        uintCV(4242323297402406),
        tupleCV({
            'for-auction': falseCV(),
            'for-sale': falseCV(),
            'price': uintCV(0),
        }),
        tupleCV({
            name: stringAsciiCV("Test 1"),
            metadata: stringAsciiCV("https://www.example.com/nft/1"),
        }),
        noneCV(),
    ],
    senderKey: userPrivateKey,
    validateWithAbi: true,
    anchorMode: anchorMode,
    network: stacksNetwork,
    sponsored: true,
};
const transaction = await makeContractCall(txOptions);
const serializedTx = transaction.serialize().toString('hex');

//Deserialize & sponsor
const bufferReader = new BufferReader(Buffer.from(serializedTx, 'hex'));
const deserializedTx = deserializeTransaction(bufferReader);
const fee = 1000n;
const sponsorOptions = {
    transaction: deserializedTx,
    sponsorPrivateKey: sponsorPrivateKey,
    fee,
};
const sponsoredTx = await sponsorTransaction(sponsorOptions);
const broadcastResponse = await broadcastTransaction(sponsoredTx, stacksNetwork);
console.log(JSON.stringify(broadcastResponse));
```

**Expected behavior**
The transaction is sponsored & processed. Removing the `sponsored: true` and broadcasting the first transaction works as expected. Adding the sponsor is where this issue comes up.

**Console log**
If applicable, add console log to help explain your problem.

**Desktop (please complete the following information):**
 - OS: MacOS 11.5.2
 - Browser: N/A
 - Node: v14.18.1

**Additional context**
This issue only occurs for sponsored transactions. Generating new keys (`yarn stx make_keychain -t`) does not solve this issue, and I validated both keys used are valid via `yarn stx get_address PRIVATE_KEY_HERE -t`.

**Comment by ahsan-javaid** [comment]
The `Invalid signature` message is coming from `stacks-blockchain` due to `SignatureValidation` failed. [Here is possible list of errors](https://github.com/blockstack/stacks-blockchain/blob/bf4a577113bde7b98c9bffc5e8aa09d19d8946d4/docs/rpc-endpoints.md#post-v2transactions)

Below is the snippet that works fine for sponsored transaction as per readme.
```
import {
  makeContractCall,
  AnchorMode,
  uintCV,
  stringAsciiCV,
  stringUtf8CV
} from '@stacks/transactions';
import {
  sponsorTransaction,
  BufferReader,
  deserializeTransaction,
  broadcastTransaction,
} from '@stacks/transactions';
import { StacksTestnet } from '@stacks/network';

void(async () => {
  const contractAddress = 'STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6';
  const contractName = 'ahsan_contract_cli1';
  const userPrivateKey = 'b244296d5907de9864c0b0d51f98a13c52890be0404e83f273144cd5b9960eed01';
  const sponsorPrivateKey = '21d43d2ae0da1d9d04cfcaac7d397a33733881081f0b2cd038062cf0ccbb752601';
  const stacksNetwork = new StacksTestnet();
  const anchorMode = AnchorMode.Any;

  const txOptions = {
    contractAddress: contractAddress,
    contractName: contractName,
    functionName: 'testfunc',
    functionArgs: [uintCV(100), stringAsciiCV('afa'), stringUtf8CV('aa')],
    senderKey: userPrivateKey,
    validateWithAbi: true,
    anchorMode: anchorMode,
    network: stacksNetwork,
    sponsored: true,
  };
  const transaction = await makeContractCall(txOptions);
  const serializedTx = transaction.serialize().toString('hex');

  // Deserialize & sponsor
  const bufferReader = new BufferReader(Buffer.from(serializedTx, 'hex'));
  const deserializedTx = deserializeTransaction(bufferReader);
  const fee = 400n;
  const sponsorOptions = {
    transaction: deserializedTx,
    sponsorPrivateKey: sponsorPrivateKey,
    fee,
  };
  const sponsoredTx = await sponsorTransaction(sponsorOptions);
  const broadcastResponse = await broadcastTransaction(sponsoredTx, stacksNetwork);
  console.log(JSON.stringify(broadcastResponse));

})();

```

You can copy/paste to execute the above snippet.

It might be possible some of your inputs/keys are causing this issue. It will be great if you can mention exact `snippet` with values filled to reproduce the issue. Like i provided above.
So that i can let you know which value is causing this issue exactly. Thanks

**Comment by alihamze** [comment]
I copied the code above and ran it directly and got exactly the same error:
```json
{"error":"transaction rejected","reason":"SignatureValidation","reason_data":{"message":"Invalid signature: failed to recover public key"},"txid":"9bfbd9f6df9d41b0ea41a9c2fe41e442a4a1e200a89e4c002af07398db97df32"}
```

Steps taken:
1. Copy the code above into a new file named `test.mjs` (installed dependencies list below if that helps)
2. Run with `node test.mjs`
    - `node -v`: `v14.18.1` 
3. Get the error above

<details>
<summary>Yarn installed dependencies</summary>

@stacks/* packages in package.json:
```json
"@stacks/cli": "^2.0.1",
"@stacks/common": "^2.0.1",
"@stacks/network": "^2.0.1",
"@stacks/profile": "^2.0.1",
"@stacks/transactions": "^2.0.1",
```

Result of `yarn list | grep '@stacks'`:

```
├─ @stacks/auth@1.2.3
│  ├─ @stacks/common@^1.2.2
│  ├─ @stacks/common@1.2.2
│  ├─ @stacks/encryption@^1.2.3
│  ├─ @stacks/network@^1.2.2
│  ├─ @stacks/network@1.2.2
│  │  └─ @stacks/common@^1.2.2
│  ├─ @stacks/profile@^1.2.3
│  ├─ @stacks/profile@1.2.3
│  │  ├─ @stacks/common@^1.2.2
│  │  ├─ @stacks/encryption@^1.2.3
│  │  ├─ @stacks/network@^1.2.2
├─ @stacks/blockchain-api-client@0.34.2
├─ @stacks/cli@2.0.1
│  ├─ @stacks/auth@^1.2.3
│  ├─ @stacks/blockchain-api-client@^0.34.1
│  ├─ @stacks/common@^2.0.1
│  ├─ @stacks/network@^1.2.2
│  ├─ @stacks/network@1.2.2
│  │  ├─ @stacks/common@^1.2.2
│  │  └─ @stacks/common@1.2.2
│  ├─ @stacks/stacking@^2.0.1
│  ├─ @stacks/storage@^2.0.1
│  ├─ @stacks/transactions@^2.0.1
├─ @stacks/common@2.0.1
├─ @stacks/encryption@1.2.3
│  ├─ @stacks/common@^1.2.2
│  ├─ @stacks/common@1.2.2
├─ @stacks/network@2.0.1
│  └─ @stacks/common@^2.0.1
├─ @stacks/profile@2.0.1
│  ├─ @stacks/common@^2.0.1
│  ├─ @stacks/network@^1.2.2
│  ├─ @stacks/network@1.2.2
│  │  ├─ @stacks/common@^1.2.2
│  │  └─ @stacks/common@1.2.2
│  ├─ @stacks/transactions@^2.0.1
├─ @stacks/stacking@2.0.1
│  ├─ @stacks/common@^2.0.1
│  ├─ @stacks/network@^1.2.2
│  ├─ @stacks/network@1.2.2
│  │  ├─ @stacks/common@^1.2.2
│  │  └─ @stacks/common@1.2.2
│  ├─ @stacks/stacks-blockchain-api-types@^0.61.0
│  ├─ @stacks/transactions@^2.0.1
├─ @stacks/stacks-blockchain-api-types@0.61.0
├─ @stacks/storage@2.0.1
│  ├─ @stacks/auth@^1.2.3
│  ├─ @stacks/common@^2.0.1
│  ├─ @stacks/encryption@^2.0.1
│  ├─ @stacks/encryption@2.0.1
│  │  ├─ @stacks/common@^2.0.1
├─ @stacks/transactions@2.0.1
│  ├─ @stacks/common@^2.0.1
│  ├─ @stacks/network@^1.2.2
│  ├─ @stacks/network@1.2.2
│  │  ├─ @stacks/common@^1.2.2
│  │  └─ @stacks/common@1.2.2
```
</details>

As for the inputs, here are the values I left out (just replace the first 4 lines of my original snippet with these since the rest is the same):

```js
const contractAddress = "ST2124A54ZRRE8TCK86RYSSSNNX9QNQHFNA8SQH15";
const contractName = "fellow-white-walrus";
const userPrivateKey = "909ff759a9460d841d8bc6250bbd59ab52701e204c3810cade289a356f8e9dfc01";
const sponsorPrivateKey = "568d7580a7cbaaf1159adc484687c58d29714e6e8adaf1435b19cb474adf036101";
```

**Comment by ahsan-javaid** [comment]
Got it. This is already fixed in this pr: https://github.com/blockstack/stacks.js/pull/1103

I tested it in master branch that's why not reproduced on my side because the fix for this is already merged. 

@reedrosenbluth Need to cut the release on npm so that users can get the fixed version of `sponsorTransaction `

**Comment by alihamze** [comment]
I can confirm that it's fixed on the master branch, is there an ETA on when a new release will be available?

**Comment by diwakergupta** [comment]
Fixed with #1103 , will be available in the next release.

**Comment by diwakergupta** [comment]
Fixed with #1103 , will be available in the next release.
